climada/engine/supplychain.py

```diff
@@ -0,0 +1,395 @@
+        Automotive Production", Review of International Economics., 23: 575–605
+
+        The function supports WIOD tables release available here:
+            http://www.wiod.org/database/wiots16
+
+        Parameters:
+            year (int): Year of wiot table. Valid years go from 2000 to 2014.
+                        Default 2014.
+            file_path (str): Path to folder where the wiod table is stored.
+                            Deafult is SYSTEM_DIR. If data are not present, they
+                            will be downloaded in save_dir, i.e., ~/climada/data/results
+                            If user-defined, user_data must be set to True
+            user_data (bool): If user data are provided, Default False
+            rows_range (tuple): Start and end position of rows with data
+            col_iso3 (int): pos of col with iso3 countries
+            cols_data_range (tuple): start and end position of cols with data
+            cols_sect_range (tuple): start and end position of sectors' names
+
+            Default values of the last four args allow reading the full wiot table.
+        """
+
+        file_name = 'WIOT{}_Nov16_ROW.xlsb'.format(year)
+        results_folder = Path(file_path).joinpath('results')
+
+        if not user_data:
+            try:
+                # if results folder exists but file does not - > download file
+                if file_name not in os.listdir(results_folder):
+                    os.chdir(SYSTEM_DIR)
+                    download_link = WIOD_FILE_LINK + file_name
+                    download_file(download_link)
+                    LOGGER.debug('Downloading WIOT table for year %s', year)
+
+                    mriot = pd.read_excel(os.path.join(results_folder, file_name),
+                                  engine='pyxlsb')
+
+                # if folder and file exist -> load file
+                else:
+                    mriot = pd.read_excel(os.path.join(results_folder, file_name),
+                                  engine='pyxlsb')
+
+            # if folder does not exist -> create folder and download file
+            except FileNotFoundError:
+                os.chdir(SYSTEM_DIR)
+                download_link = WIOD_FILE_LINK + file_name
+                download_file(download_link)
+                LOGGER.debug('Downloading WIOT table for year %s', year)
+
+                mriot = pd.read_excel(os.path.join(results_folder, file_name),
+                                  engine='pyxlsb')
+        else:
+            mriot = pd.read_excel(os.path.join(file_path, file_name),
+                                  engine='pyxlsb')
+
+        row_st,row_end=rows_range
+        col_sect,row_sect_end=cols_sect_range
+        col_data_st,col_data_end=cols_data_range
+
+        sectors = mriot.iloc[row_st:row_sect_end, col_sect].values
+        countries_iso3 = np.unique(mriot.iloc[row_st:row_end, col_iso3])
+        # move Rest Of World (ROW) at the end of the array as countries
```

Comment:
ROW is a "country" of the WIOT table - it is assumed whoever uses the module knows what the WIOT table is and how it is structured otherwise it is hard to know what the module does..

Reply:
aaa ok, I misundertood that, because below in the code it says that the `ROW` can be multiples that need to be aggregated

>           # Sum needed below in case of many ROWs, which need be aggregated.
            # Another option is to keep them separate, but it requires changing
            # all functions below as impact matrix will no longer have
            # dim = (n_years, self.n_countries*self.n_sectors) but
            # dim = (n_years, (self.n_countries+#ROWs_countries-1)*self.n_sectors)

Reply:
Yes. So basically you have these bunch of countries from the WIOT table, plus one which is ROW. Then, for to the countries you have exposure and impact for, you must check whether this is explicitely included in the WIOT or it falls in ROW..

Reply:
I see. It would probably be simpler to compute the impact for all the `ROW` at once then no?

Reply:
You need to iterate to known which one is ROW and which one is not - in the meantime you can assess their damage

Reply:
Why? You can select all the exposures at once that are in the `ROW`, and then compute the impact from that...

Reply:
Ok, so what happens here is that all countries which are NOT explicitely modeled by the WIOD table ends up in ROW. Therefore, in case one is dealing with more than a country which ends up in ROW, their overall impact is summed up so to give the ROW "country" overall impact. This guarantees consistency with the dimensions of the WIOD table. In general, I do not think it makes sense modelling _any_ country which ends up in ROW, as that is a very crude approximation of their IO structure. The comment below has been reduced as I found it myself quite consusing (or, to the least, providing unnecessary info)

```
      # Sum needed below in case of many ROWs, which need be aggregated.
       # Another option is to keep them separate, but it requires changing
       # all functions below as impact matrix will no longer have 
      # dim = (n_years, self.n_countries*self.n_sectors) but
       # dim = (n_years, (self.n_countries+#ROWs_countries-1)*self.n_sectors)
```